lib/libcgroup-3.1.0/doc/man/lssubsys.1
.\" Copyright (C) 2009 Red Hat, Inc. All Rights Reserved.
.\" Written by Ivana Hutarova Varekova <varekova@redhat.com>

.TH LSSUBSYS  1 2009-11-05 "Linux" "libcgroup Manual"
.SH NAME

lssubsys \- list hierarchies containing given subsystem

.SH SYNOPSIS
\fBlssubsys\fR  [\fB-m|--mount-points\fR] [\fB-M|--all-mount-points\fR]
[\fIcontroller\fR] [...]
.br
\fBlssubsys\fR  [\fB-m|--mount-points\fR] [\fB-M|--all-mount-points\fR]
[\fB-i|--hierarchies\fR] [\fB-a|--all\fR]
.br
\fBlssubsys\fR  [\fB-h|--help\fR]

.SH DESCRIPTION
\fBlssubsys\fR list the hierarchies containing
the given subsystem(s). If no subsystem is defined, it
displays all mounted hierarchies.

.TP
.B controller
defines the subsystem whose hierarchy is shown.
If this parameter is not used, the command will
list all mounted subsystems.

.TP
.B -m, --mount-points
Display mount points. Only the first mount points of shown hierarchies are displayed.

.TP
.B -M, --all-mount-points
Display mount points. All mount points of shown hierarchies are displayed, each on
separate line.

.TP
.B -a, --all
Display all subsystems (including unmounted ones).

.TP
.B -i, --hierarchies
Display the attached hierarchy number if the subsystem is in a hierarchy.
If the option
.B -m
is also used, the hierarchy number is shown only for unmounted hierarchies.

.TP
.B -h, --help
Display help and exit.

.SH EXAMPLE
The command
.B lssubsys -am
lists all subsystems which are present in the system,
mounted ones will be shown with their mount point:
.RS
.nf
$ lssubsys -am
ns
cpuset
cpu,devices
cpuacct /cgroup/cpuacct
memory /cgroup/memory
freezer /cgroup/freezer
net_cls /cgroup/net_cls
.fi
.RE
.LP
The command
.B lssubsys -ami
lists all subsystems which are present in the system,
mounted subsystems are listed with their mount points,
unmounted ones which are in a hierarchy are listed with their hierarchy number:
.RS
.nf
$ lssubsys -ami
ns
cpuset 5
cpu,devices 6
cpuacct /cgroup/cpuacct
memory /cgroup/memory
freezer /cgroup/freezer
net_cls /cgroup/net_cls
.fi

.SH NOTES
.TP
lssubsys is currently supported on cgroups v1 only.

.SH ENVIRONMENT VARIABLES
.TP
.B CGROUP_LOGLEVEL
controls verbosity of the tool. Allowed values are \fBDEBUG\fR,
\fBINFO\fR, \fBWARNING\fR or \fBERROR\fR.

.RE
.SH SEE ALSO
lscgroup (1), cgcreate (1), cgdelete (1),
cgconfig.conf (5)


ChangeLog
DOMjudge Programming Contest Judging System

Version 8.4.0DEV
---------------------------

Version 8.3.0 - 31 May 2024
---------------------------
 - [security] Close metadata file descriptor for the child in runguard.
 - Document that minimum PHP version is now 8.1.0.
 - Upgrade Symfony to 6.2 and upgrade other library dependencies.
 - Show link to problem text on jury problem overview and contest detail pages.
 - Add label to teams to be in sync with ICPC CCS spec.
 - Add feedback/teammessage.txt for automated teamfeedback, see: #2016.
 - Clarifications can now use markdown syntax.
 - Upgrade Bootstrap to 5.2 and dependent components.
 - Bug fixed to correctly select current user during configuration of installation.
 - Properly fix removal of problem/submission data.
 - Filter away balloons of jury members in balloon runner interface.
 - Fix upgrade path for executables, see: #1826.
 - Collect compiler and runner versions before compilation, useful to compare
   judgehosts.
 - Various improvements to shadow config used at ICPC World Finals.
 - Show the historic timelimit which was used when judging the submission.
 - Fixes to the documentation and import of problems with @EXPECTED_RESULTS@
   tags.
 - Allow filtering on affilition in submission page for jury for multisite
   contests.
 - Add preview of the contest problemset to the jury page before contest
   start.
 - Add optional Adminer for easy database access.
 - Fix medal awards when skipping categories.
 - Add direct button to public page for registering when enabled.
 - Scale batch size of judge tasks dynamically.
 - Rename room to location for teams.
 - Add option to upload contest problemset document.
 - Improve on the advice for rejuding on config changes, only when relevant
   advice to start a rejudging, similar for updating the scoreboard cache.
 - For interactive problems: add explicit statement to when one of sides of
   the interactions ran closed their output.
 - Allow importing a (generic) contest warning which is displayed on the scoreboard.
 - Added option in the dj_setup_database to make dumps, useful before major
   upgrades and during the contest before problem imports.
 - Fixed some bugs to become OpenAPI spec compliant again.
 - Prevent a race condition leading to two submissions receiving the same
   colour balloon.
 - Compile executables with `-std=gnu++20` now.
 - Record the testcase directory for judging runs, this makes investigating
   easier.
 - Set default C/C++ standards in configure script, we intend to ship all
   scripts free of warnings/errors.
 - Respect basic authentication in the URL for the import-contest when .netrc
   exists.
 - Fix CORS header for API to make extending with other tools easier.
 - Fix eventfeed when SQL commits happen out of order to send all events.
 - Lazy load the team pictures for the scoreboard, to prevent spamming access
   log with HTTP404s
 - Make TLE detection more robust for interactive problems.
 - Added option for ICPC online scoreboard with no rank displays.

Version 8.2.0 - 6 March 2023
----------------------------
 - Various bugfixes & UI improvements.
 - Add support for Rust as submission language.
 - Add teamphotos to scoreboard zip.
 - Add jury option to to download ZIP with all problem samples, attachments
   and problem statement.
 - Extend team import with optional room and other API extensions, see the API
   doc for further details.
 - Drop support for shadowing a single contest from multiple sources.
 - Log account logins via API separately.
 - Add support for PHP v8.2.
 - Upgraded library dependencies.
 - Removed output of detected entry point to teams.
 - Added option to add custom apt repositories to judge chroot.
 - Added CLI script to configure the domserver with a yaml file.
 - Changed activation behaviour of contests imported via UI to be directly
   available.
 - New option added to reset login status of teams between contests.
 - Highlight submissions which were judged with compiler flags which since
   have changed. Useful when you find wrong flags during the contest.
 - Handle network errors for judgedaemon when domserver is unavailable.
 - Added super lazy mode useful for ICPC analyst instances. Here nothing is
   judged unless manually requested.
 - Split problem import messages in info/warning/danger, to make sure errors
   are not missed when importing.
 - Store user login sessions in database.
 - Contest import CLI script is now rewritten in python.
 - The example test problems are now imported via the contest package format.

Version 8.1.0 - 1 June 2022
---------------------------
 - Document that minimum PHP version is now 7.4.0.
 - Add support for PHP 8.1.
 - Upgrade Symfony to 5.4 LTS and upgrade other library dependencies.
 - Replace calls to deprecated `strftime` with `date`. Note: if you have
   configured a custom `time_format`, you need to update it to the format
   supported by `date`.
 - Dropped the presentation-error verdict, databases from before DOMjudge 5.0
   will be altered to replace the verdict with wrong-answer.
 - Changed API import to give admin users a team.
 - Balloon users can now be created via the API.
 - Shadow mode will now fix missing data by reading from the primary CCS.
 - Fixed cgroup naming for docker to prevent sharing the same container.
 - Removed python2 in the chroot, remove python3 in favor of pypy3 in chroot.
 - Added support for sentry for logging.
 - Added default external IDs on all places.
 - Fixed medals on scoreboard for filtered scoreboards.
 - Dropped old balloons CLI.
 - Reimplementation of runpipe for interactive problems.
 - Individual files can now be removed from executables.
 - Added mysql socket option, for when port/IP is not used.
 - Added support for accounts.{yaml,json} for import script.
 - Added support for judgeremaining for problems.
 - Add option to delete existing testcases/attachments/other data when
   importing a problem.

lib/libcgroup-3.1.0/src/tools/lssubsys.c
// SPDX-License-Identifier: LGPL-2.1-only
/**
 * Copyright (C) 2009 Red Hat, Inc. All Rights Reserved.
 * Written by Ivana Hutarova Varekova <varekova@redhat.com>
 */

#include "tools-common.h"

#include <libcgroup.h>
#include <libcgroup-internal.h>

#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <stdio.h>


enum flag {
	FL_MOUNT = 1,		/* show the mount points */
	FL_LIST = 2,
	FL_ALL = 4,		/* show all subsystems - not mounted too */
	FL_HIERARCHY = 8,	/* show info about hierarchies */
	FL_MOUNT_ALL = 16	/* show all mount points of hierarchies */
};

typedef char cont_name_t[FILENAME_MAX];

static void usage(int status, const char *program_name)
{
	if (status != 0) {
		err("Wrong input parameters, try %s -h' for more information.\n", program_name);
		return;
	}

	info("Usage: %s [-i] [-m] [-M] [controller] [...]\n", program_name);
	info("   or: %s [-a] [-i] [-m] [-M]\n", program_name);
	info("List information about given controller(s) If no controller ");
	info("is set list information about all mounted controllers.\n");
	info("  -a, --all			Display information ");
	info("about all controllers (including not mounted ones)\n");
	info("  -h, --help			Display this help\n");
	info("  -i, --hierarchies		Display information about hierarchies\n");
	info("  -m, --mount-points		Display mount points\n");
	info("  -M, --all-mount-points	Display all mount points\n");
	info("(Note: currently supported on cgroups v1 only)\n");
}

static int print_controller_mount(const char *controller, int flags, cont_name_t cont_names,
				  int hierarchy)
{
	char path[FILENAME_MAX];
	void *handle;
	int ret = 0;

	/* print only hierarchy name */
	if (!(flags & FL_MOUNT) && !(flags & FL_HIERARCHY))
		info("%s\n", cont_names);

	/* print only hierarchy name and number*/
	if (!(flags & FL_MOUNT) && (flags & FL_HIERARCHY))
		info("%s %d\n", cont_names, hierarchy);

	if (flags & FL_MOUNT) {
		/* print hierarchy name and mount point(s) */
		ret = cgroup_get_subsys_mount_point_begin(controller, &handle, path);
		/* intentionally ignore error from above call */
		while (ret == 0) {
			info("%s %s\n", cont_names, path);
			/* first mount record is enough */
			if (!(flags & FL_MOUNT_ALL))
				goto stop;
			ret = cgroup_get_subsys_mount_point_next(&handle, path);
		}
		if (ret == ECGEOF)
			ret = 0;
stop:
		cgroup_get_subsys_mount_point_end(&handle);
	}
	return ret;
}

/* display all controllers attached to the given hierarchy */
static int print_all_controllers_in_hierarchy(const char *tname, int hierarchy, int flags)
{
	struct controller_data info;
	enum cg_version_t version;
	cont_name_t cont_names = "\0";
	cont_name_t cont_name;
	int first = 1;
	void *handle;
	int ret = 0;

	/*
	 * Initialize libcgroup and intentionally ignore its result, no
	 * mounted controller is valid use case.
	 */
	(void) cgroup_init();

	ret = cgroup_get_all_controller_begin(&handle, &info);
	if ((ret != 0) && (ret != ECGEOF)) {
		err("cannot read controller data: %s\n", cgroup_strerror(ret));
		return ret;
	}

	while (ret != ECGEOF) {
		ret = cgroup_get_controller_version(info.name, &version);
		if (ret)
			goto end;

		/*
		 * v1 controllers should be in the hierachy.
		 * v2 controllers will have a hierarchy value of zero
		 */
		if (version == CGROUP_V1 && info.hierarchy != hierarchy)
			goto next;

		if (first) {
			/* the first controller in the hierarchy */
			strncpy(cont_name, info.name, FILENAME_MAX-1);
			cont_name[sizeof(cont_name) - 1] = '\0';

			strncpy(cont_names, info.name, FILENAME_MAX-1);
			cont_names[sizeof(cont_names) - 1] = '\0';
			first = 0;
		} else {
			/* the next controller in the hierarchy */
			strncat(cont_names, ",", FILENAME_MAX-1);
			strncat(cont_names, info.name, FILENAME_MAX-1);
			cont_names[sizeof(cont_names) - 1] = '\0';
		}

next:
		ret = cgroup_get_all_controller_next(&handle, &info);
		if (ret && ret != ECGEOF)
			goto end;
	}

	ret = print_controller_mount(cont_name, flags, cont_names, hierarchy);
end:
	cgroup_get_all_controller_end(&handle);

	if (ret == ECGEOF)
		ret = 0;

	return ret;
}


/*
 * go through the list of all controllers gather them based on hierarchy
 * number and print them
 */
static int cgroup_list_all_controllers(const char *tname, cont_name_t cont_name[CG_CONTROLLER_MAX],
				       int c_number, int flags)
{
	struct controller_data info;
	int h_list[CG_CONTROLLER_MAX];	/* list of hierarchies */
	int is_on_list = 0;
	int counter = 0;
	void *handle;
	int ret;
	int j;

	ret = cgroup_get_all_controller_begin(&handle, &info);
	while (ret == 0) {
		if (info.hierarchy == 0) {
			/* the controller is not attached to any hierachy */
			if (flags & FL_ALL) {
				/* display only if -a flag is set */
				info("%s\n", info.name);
			}
		}
		is_on_list = 0;
		j = 0;
		while ((is_on_list == 0) && (j < c_number)) {
			if (strcmp(info.name, cont_name[j]) == 0) {
				is_on_list = 1;
				break;
			}
			j++;
		}

		if ((info.hierarchy != 0) && ((flags & FL_ALL) ||
		    (!(flags & FL_LIST) || (is_on_list == 1)))) {
			/*
			 * the controller is attached to some hierarchy
			 * and either should be output all controllers,
			 * or the controller is on the output list
			 */
			h_list[counter] = info.hierarchy;
			counter++;
			for (j = 0; j < counter-1; j++) {
				/*
				 * the hierarchy already was on the list
				 * so remove the new record
				 */
				if (h_list[j] == info.hierarchy) {
					counter--;
					break;
				}
			}
		}

		ret = cgroup_get_all_controller_next(&handle, &info);
	}
	cgroup_get_all_controller_end(&handle);
	if (ret == ECGEOF)
		ret = 0;
	if (ret) {
		err("cgroup_get_controller_begin/next failed (%s)\n", cgroup_strerror(ret));
		return ret;
	}

	for (j = 0; j < counter; j++)
		ret = print_all_controllers_in_hierarchy(tname,	h_list[j], flags);

	return ret;
}

int main(int argc, char *argv[])
{
	static struct option options[] = {
		{"help",		0, 0, 'h'},
		{"mount-points",	0, 0, 'm'},
		{"all-mount-points",	0, 0, 'M'},
		{"all",			0, 0, 'a'},
		{"hierarchies",		0, 0, 'i'},
		{0, 0, 0, 0}
	};

	cont_name_t cont_name[CG_CONTROLLER_MAX];
	int c_number = 0;
	int flags = 0;
	int ret = 0;
	int c;
	int i;

	for (i = 0; i < CG_CONTROLLER_MAX; i++)
		cont_name[i][0] = '\0';

	/* parse arguments */
	while ((c = getopt_long(argc, argv, "mMhai", options, NULL)) > 0) {
		switch (c) {
		case 'h':
			usage(0, argv[0]);
			return 0;
		case 'm':
			flags |= FL_MOUNT;
			break;
		case 'M':
			flags |= FL_MOUNT | FL_MOUNT_ALL;
			break;
		case 'a':
			flags |= FL_ALL;
			break;
		case 'i':
			flags |= FL_HIERARCHY;
			break;
		default:
			usage(1, argv[0]);
			exit(EXIT_BADARGS);
		}
	}

	/* read the list of controllers */
	while (optind < argc) {
		flags |= FL_LIST;
		strncpy(cont_name[c_number], argv[optind], FILENAME_MAX);
		cont_name[c_number][FILENAME_MAX-1] = '\0';
		c_number++;
		optind++;
		if (optind == CG_CONTROLLER_MAX) {
			err("Warning: too many parameters\n");
			break;
		}
	}

	ret = cgroup_list_all_controllers(argv[0], cont_name, c_number, flags);

	return ret;
}
